tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -1093,6 +1122,53 @@ private[optimizer] abstract class OptimizerCore(
       keepOnlySideEffects(tree)
   }
 
+  /** Finishes the bindings and statements followed by a result to get a
+   *  normal [[Tree]].
+   *  This method must not be called more than once per `BindingOrStat` and
+   *  per translation.
+   *  By "per translation", we mean in an alternative path through
+   *  `tryOrRollback`. It could still be called several times as long as
+   *  it is once in the 'try' part and once in the 'fallback' part.
+   */
+  private def finishTransformBindings(bindingsAndStats: List[BindingOrStat],
+      result: Tree): Tree = {
+    bindingsAndStats.foldRight(result) {
+      case (Left(PreTransBinding(localDef, value)), innerBody) =>
+        implicit val pos = value.pos
+
+        val LocalDef(tpe, mutable, replacement) = localDef
+
+        val (name, originalName, used) = (replacement: @unchecked) match {
+          case ReplaceWithVarRef(name, originalName, used, _) =>
+            (name, originalName, used)
+          case ReplaceWithRecordVarRef(name, originalName, _, used, _) =>
+            (name, originalName, used)
+        }
+
+        if (used.value) {
+          val ident = Ident(name, originalName)
+          val varDef = resolveLocalDef(value) match {
+            case PreTransRecordTree(valueTree, valueTpe, cancelFun) =>
+              val recordType = valueTree.tpe.asInstanceOf[RecordType]
+              if (!isImmutableType(recordType))
```

Comment:
Yes and no. Even if you use the `VarDef` only once, maybe you're *also* still using the rhs it was copied from. Consider (pseudo-scala notation):
```scala
val rec1 = { val x: Int = 5; var y: Int = 5 }
println(rec1.y) // prints 5
val rec2 = rec1
rec2.y = 3
println(rec1.y) // still prints 5
```
That happens even though I used `rec2` only once.
Whereas, if `rec1` and `rec2` had remained reference types, the last line would print 3 as well. So copying a reference and copying a mutual record value do not have the same semantics. That's why we have to abort.

To not abort here, we would have to ensure/check some kind of ownership transfer, by which `rec1` gets invalidated from then on. But that depends on the evaluation order, and checking evaluation order-dependent properties is something our IR is not good at.